flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecSink.java

```diff
@@ -220,12 +224,14 @@ private int deriveSinkParallelism(
      * messages.
      */
     private Transformation<RowData> applyKeyBy(
+            PlannerBase planner,
             Transformation<RowData> inputTransform,
             int[] primaryKeys,
             int sinkParallelism,
             boolean upsertMaterialize) {
         final int inputParallelism = inputTransform.getParallelism();
-        if ((inputParallelism == sinkParallelism || changelogMode.containsOnly(RowKind.INSERT))
+        if ((inputParallelism == sinkParallelism
+                        || getOutputChangelogMode(planner).containsOnly(RowKind.INSERT))
```

Comment:
can't we just reuse the already discovered table sink in `createSinkTransformation` and avoid discovery and creation multiple times?

Reply:
Yes, good catch.